calico/felix/fetcd.py

```diff
@@ -179,6 +191,75 @@ def _periodically_report_status(self):
                 sleep_time = interval + jitter
                 gevent.sleep(sleep_time)
 
+    @logging_exceptions
+    def _report_endpoint_status(self):
+        _log.info("Started per-endpoint status reporting thread. Waiting for "
+                  "config.")
+        self._watcher.configured.wait()
+        if not self._config.REPORT_ENDPOINT_STATUS:
+            _log.info("Endpoint status reporting disabled.")
+            return
+        while True:
+            # Wait for some work to do, then clear the flag ready for next
+            # time.  There is no race here with someone else setting the flag
+            # again because the flag is only set after the shared data
+            # structure is updated and we don't read it until after we've
+            # cleared the flag.
+            self._endpoint_status_work_to_do.wait()
+            self._endpoint_status_work_to_do.clear()
+            with self._endpoint_status_lock:
+                # This lock is probably not required in gevent due to lack of
+                # yield points in this code.  It is required in normal python,
+                # since, otherwise, _mark_endpoint_dirty could dereference
+                # self._dirty_endpoints, then get de-scheduled and
+                # subsequently write to the set after we start iterating over
+                # it.
+                dirty_endpoints = self._dirty_endpoints
+                self._dirty_endpoints = set()
+            for ep_id in dirty_endpoints:
+                status = self._endpoint_status.get(ep_id)
+                try:
+                    self.write_endpoint_status_to_etcd(ep_id, status)
+                except EtcdException:
+                    _log.error("Failed to report status for %s, will retry",
+                               ep_id)
+                    # Add it into the next dirty set.  Retrying in the next
+                    # batch ensures that we try to update all of the dirty
+                    # endpoints before we do any retries, ensuring fairness.
+                    self.mark_endpoint_dirty(ep_id)
+                # Rate-limit our writes, with jitter.
+                gevent.sleep(self._config.ENDPOINT_REPORT_DELAY *
+                             (0.9 + (random.random() * 0.2)))
+
+    def mark_endpoint_dirty(self, ep_id):
+        if not self._config.REPORT_ENDPOINT_STATUS:
+            _log.debug("Endpoint status reporting disabled, ignoring.")
+            return
+        _log.debug("Marking endpoint %s dirty, needs resync", ep_id)
+        with self._endpoint_status_lock:
+            self._dirty_endpoints.add(ep_id)
+            self._endpoint_status_work_to_do.set()
+
+    def write_endpoint_status_to_etcd(self, ep_id, status):
```

Comment:
Should this check for REPORT_ENDPOINT_STATUS? If not (I think the caller has checked) I would quite like an assert here. Because asserts make me feel safe.